controllers/kubeadmconfig_controller.go

```diff
@@ -108,16 +122,29 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 		return ctrl.Result{}, err
 	}
 
-	// maybe do something with cluster some day
-	// maybe do something interesting here some day
-	config.Status.BootstrapData = []byte("hello world")
-	config.Status.Ready = true
+	// Generate the cloud-init for controlplane init
+	if r.BootstrapDataGenerator == nil {
+		r.BootstrapDataGenerator = &BootstrapDataGenerator{
+			Logger: r.Log.WithName("user-data-generator"),
+		}
+	}
+
+	if util.IsControlPlaneMachine(machine) {
```

Comment:
Do we want to do this, or just check to see if `config` has the `ClusterConfiguration` and `InitConfiguration` fields set? Or both?

Reply:
Switching to checking for ClusterConfiguration/InitConfiguration would potentially remove the need for a lock completely.

Reply:
We'd still need it. If you submit multiple control plane machines, or worker machines, everything except the init has to wait.

Reply:
Yes, the other Machines need to wait, but that doesn't necessitate a lock, only that they wait for the "Control Plane Ready" annotation. The lock was only needed for determining which control plane Machine should do the init vs join.

Reply:
Fair point

Reply:
Hold up, lemme try to rephrase to check my understanding:

There is only one reconciler running at a time and it is single threaded so we'll reconcile machines one at a time. This makes locking pretty much already not necessary.

The first control plane machine that comes in gets the clusterconfiguration/initconfiguration and annotation is added to some object (?) "control-plane-ready: yes". Every following control-plane machine that gets reconciled checks this annotation and will be a control-plane join. Workers are always joins.

I'm trying to figure out what object to store this annotation on and if "control-plane-ready" is the right name for this. It doesn't feel right, unless perhaps the value becomes a state (no => no control plane has been created so the current machine can be the first; pending => a control plane has been assigned and is being init'd => yes the control plane is up and running)

Reply:
> There is only one reconciler running at a time and it is single threaded so we'll reconcile machines one at a time. This makes locking pretty much already not necessary.

We shouldn't assume this is always going to be the case.

Also, with CAPA & v1alpha1, it can finish reconciling a Machine and move on to the next Machine before the control plane is ready, so we need this in v1alpha1, even with it being single-threaded.

> The first control plane machine that comes in gets the clusterconfiguration/initconfiguration

This would be manually configured by the user.

> annotation is added to some object (?) "control-plane-ready: yes".

For CAPA v1alpha1, we put this on the `Cluster`.

> Every following control-plane machine that gets reconciled checks this annotation and will be a control-plane join. Workers are always joins.

The user would be expected to only fill in `KubeadmConfig.Spec.JoinConfiguration` for these.

> I'm trying to figure out what object to store this annotation on and if "control-plane-ready" is the right name for this. It doesn't feel right, unless perhaps the value becomes a state (no => no control plane has been created so the current machine can be the first; pending => a control plane has been assigned and is being init'd => yes the control plane is up and running)

If you want an alternative that doesn't require an annotation, we could list all the Machines for the Cluster, and if at least one of them has status.nodeRef set, that means the control plane is ready (the nodeRef is only set if we can reach the apiserver).

Reply:
Ahh of course the user sets it. yeah, that absolutely works for me.